lib/gc/Transforms/GPU/LinalgToXeGPU.cpp

```diff
@@ -686,53 +720,214 @@ createDescriptorTiles(PatternRewriter &rewriter, Location loc, Value src,
   return tiles;
 }
 
-// Create coarse sub-tiles to be loaded by the current subgroup.
+// Split a source into a series of 1D descriptor tiles. Each descriptor tile
+// loads exactly 32 elements.
 //
-// The shape to be loaded is split into the largest 2D loads supported
-// by the hardware.
+// The descriptors collectively load blocks of the 'loadShape2D' shape
+// with the chunk sizes being 'tileSize2D'.
 //
-// The load subgroup tiles are ordered in row-major fashion with respect to the
-// source shape.
-static SmallVector<Value> createCoarseDscTiles(PatternRewriter &rewriter,
-                                               Location loc, Value src,
-                                               ArrayRef<int64_t> sgTile,
-                                               bool isVnni,
-                                               bool transpose = false) {
-  assert(sgTile.size() <= 2 &&
+// The descriptor sub-tiles are ordered in row-major fashion with respect to the
+// whole load tile.
+static SmallVector<Value> createScatterDescriptorTiles(
+    PatternRewriter &rewriter, Location loc, Value flatMemref,
+    ArrayRef<int64_t> loadShape2D, ArrayRef<int64_t> tileSize2D,
+    ArrayRef<int64_t> memrefStrides, Value blockOffset) {
+  int64_t maxLoadSize = 32;
+
+  assert(memrefStrides.size() == 2 && "Strides must be 2D");
+  assert(memrefStrides[1] == 1 && "Only row-major strides are supported");
+  assert(loadShape2D.size() == 2 && "Load shape must be 2D");
+  assert(loadShape2D[0] * loadShape2D[1] % maxLoadSize == 0 &&
+         "Load shape must be divisible by max load size");
+  assert(tileSize2D.size() == 2 && "Descriptor tile must be 2D");
+  assert(maxLoadSize % tileSize2D[1] == 0 &&
+         "Descriptor tile must be divisible by max load size");
+
+  int64_t numLoadsPerTile = tileSize2D[0] * tileSize2D[1] / maxLoadSize;
+  // This indicates how many rows of a single tile (defined by tileSize2D) are
+  // loaded per single load operation (single load loads exactly 32 elements).
+  int64_t rowsPerLoad = maxLoadSize / tileSize2D[1];
+  int64_t numColTiles = loadShape2D[1] / tileSize2D[1];
```

Comment:
Could loadShape2D[1]  be less than tileSize2D[1]? If yes, you will get zero here and division by zero below.